git_data_model.sql

```diff
@@ -1,160 +1,337 @@
 CREATE TABLE users (
-user_id BIGSERIAL PRIMARY KEY NOT NULL,
-user_name VARCHAR (255) NOT NULL,
+id BIGSERIAL PRIMARY KEY NOT NULL,
+name VARCHAR (255) NOT NULL,
 email VARCHAR (255) NOT NULL,
 password VARCHAR(255) NOT NULL,
-UNIQUE (user_id)
+created_at TIMESTAMP WITHOUT TIME ZONE DEFAULT(NOW() AT TIME ZONE 'UTC'),
+UNIQUE (email)
 );
 
 
-INSERT INTO users (user_name, email, password)
+INSERT INTO users (name, email, password)
 VALUES ('salman', 'sal@gmail.com','pass1');
 
-INSERT INTO users (user_name, email, password)
+INSERT INTO users (name, email, password)
 VALUES ('salman2', 'sal2@gmail.com','pass2');
 
-INSERT INTO users (user_name, email, password)
+INSERT INTO users (name, email, password)
 VALUES ('salman3', 'sal3@gmail.com','pass3');
 
 SELECT * FROM users;
 
-CREATE TABLE repo (
-repo_id BIGSERIAL PRIMARY KEY NOT NULL,
-repo_name VARCHAR (255) NOT NULL,
-descriptions VARCHAR (255) NOT NULL,
-owner_id  BIGINT REFERENCES users(user_id),
-files text ,
-UNIQUE (repo_id)
+CREATE TABLE repos (
+id BIGSERIAL PRIMARY KEY NOT NULL,
+name VARCHAR (255) NOT NULL,
+description VARCHAR (255) NOT NULL,
+created_at TIMESTAMP WITHOUT TIME ZONE DEFAULT(NOW() AT TIME ZONE 'UTC'),
+updated_at TIMESTAMP WITHOUT TIME ZONE DEFAULT(NOW() AT TIME ZONE 'UTC'),
+visibility VARCHAR(255) NOT NULL
 );
 
-INSERT INTO repo (repo_name,descriptions, owner_id,files)
-VALUES ('repo1','first repo of sal', 1, 'file1');
+INSERT INTO repos (name,description,visibility)
+VALUES ('repo1','first repo of sal', 'private');
 
-SELECT * FROM repo;
 
-CREATE TABLE collaborators(
-  repo_id BIGINT,
-  collaborator_id BIGINT,
-  FOREIGN KEY  (repo_id) REFERENCES users(user_id),
-  FOREIGN KEY  (collaborator_id) REFERENCES users(user_id)
+INSERT INTO repos (name,description,visibility)
+VALUES ('repo2','second repo of sal', 'public');
+
+SELECT * FROM repos;
+
+CREATE TABLE user_repos(
+  repo_id BIGINT NOT NULL,
+  user_id BIGINT NOT NULL,
+  access_level VARCHAR(255) NOT NULL,
+  FOREIGN KEY  (repo_id) REFERENCES repos(id),
+  FOREIGN KEY  (user_id) REFERENCES users(id)
 );
 
-INSERT INTO collaborators (repo_id,collaborator_id)
-VALUES (1,2);
+CREATE INDEX index_repo_access ON user_repos(repo_id,access_level);
+
+INSERT INTO user_repos (repo_id,user_id,access_level)
+VALUES (1,1,'owner');
+
 
-INSERT INTO collaborators (repo_id,collaborator_id)
-VALUES (1,3);
+INSERT INTO user_repos (repo_id,user_id,access_level)
+VALUES (1,2,'collaborator');
 
-SELECT * FROM collaborators;
+SELECT * FROM user_repos;
 
 
 CREATE TABLE commits(
-  commit_hash VARCHAR(255) PRIMARY KEY NOT NULL,
+  hash VARCHAR(255) PRIMARY KEY NOT NULL,
```

Comment:
why not id as primary key?

Reply:
beacuse for commits every time a unique hash will be generated which will act as a primary key.so do we really have to generate another id for the same. Another thought was that since commits would not be predictable it wouldn't be guessed randomly and accessed like we would handle a transaction id in case of transactions.